File: packages/wasmvm/wasmlib/go/wasmclient/wasmclientservice.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmclient

import (
	"time"

	"github.com/iotaledger/wasp/client"
	"github.com/iotaledger/wasp/packages/cryptolib"
	"github.com/iotaledger/wasp/packages/iscp"
	"github.com/iotaledger/wasp/packages/kv/dict"
	"github.com/iotaledger/wasp/packages/subscribe"
	"github.com/iotaledger/wasp/packages/wasmvm/wasmhost"
	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib"
	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib/wasmtypes"
)

type IClientService interface {
	CallViewByHname(chainID wasmtypes.ScChainID, hContract, hFunction wasmtypes.ScHname, args []byte) ([]byte, error)
	PostRequest(chainID wasmtypes.ScChainID, hContract, hFunction wasmtypes.ScHname, args []byte, allowance *wasmlib.ScAssets, keyPair *cryptolib.KeyPair) (wasmtypes.ScRequestID, error)
	SubscribeEvents(msg chan []string, done chan bool) error
	WaitUntilRequestProcessed(chainID wasmtypes.ScChainID, reqID wasmtypes.ScRequestID, timeout time.Duration) error
}

type WasmClientService struct {
	cvt        wasmhost.WasmConvertor
	waspClient *client.WaspClient
	eventPort  string
	nonce      uint64
}

var _ IClientService = new(WasmClientService)

func NewWasmClientService(waspAPI, eventPort string) *WasmClientService {
	return &WasmClientService{waspClient: client.NewWaspClient(waspAPI), eventPort: eventPort}
}

func DefaultWasmClientService() *WasmClientService {
	return NewWasmClientService("127.0.0.1:9090", "127.0.0.1:5550")
}

func (sc *WasmClientService) CallViewByHname(chainID wasmtypes.ScChainID, hContract, hFunction wasmtypes.ScHname, args []byte) ([]byte, error) {
	iscpChainID := sc.cvt.IscpChainID(&chainID)
	iscpContract := sc.cvt.IscpHname(hContract)
	iscpFunction := sc.cvt.IscpHname(hFunction)
	params, err := dict.FromBytes(args)
	if err != nil {
		return nil, err
	}
	res, err := sc.waspClient.CallViewByHname(iscpChainID, iscpContract, iscpFunction, params)
	if err != nil {
		return nil, err
	}
	return res.Bytes(), nil
}

func (sc *WasmClientService) PostRequest(chainID wasmtypes.ScChainID, hContract, hFunction wasmtypes.ScHname, args []byte, allowance *wasmlib.ScAssets, keyPair *cryptolib.KeyPair) (reqID wasmtypes.ScRequestID, err error) {
	iscpChainID := sc.cvt.IscpChainID(&chainID)
	iscpContract := sc.cvt.IscpHname(hContract)
	iscpFunction := sc.cvt.IscpHname(hFunction)
	params, err := dict.FromBytes(args)
	if err != nil {
		return reqID, err
	}
	sc.nonce++
	req := iscp.NewOffLedgerRequest(iscpChainID, iscpContract, iscpFunction, params, sc.nonce)
	iscpAllowance := sc.cvt.IscpAllowance(allowance)
	req.WithAllowance(iscpAllowance)
	signed := req.Sign(keyPair)
	err = sc.waspClient.PostOffLedgerRequest(iscpChainID, signed)
	if err == nil {
		reqID = sc.cvt.ScRequestID(signed.ID())
	}
	return reqID, err
}

func (sc *WasmClientService) SubscribeEvents(msg chan []string, done chan bool) error {
	return subscribe.Subscribe(sc.eventPort, msg, done, false, "")
}

func (sc *WasmClientService) WaitUntilRequestProcessed(chainID wasmtypes.ScChainID, reqID wasmtypes.ScRequestID, timeout time.Duration) error {
	iscpChainID := sc.cvt.IscpChainID(&chainID)
	iscpReqID := sc.cvt.IscpRequestID(&reqID)
	_, err := sc.waspClient.WaitUntilRequestProcessed(iscpChainID, *iscpReqID, timeout)
	return err
}

File: packages/wasmvm/wasmclient/ts/wasmclientold/wasmclientservice.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as isc from "./isc"
import * as wasmlib from "wasmlib"

export interface IClientService {
    callViewByHname(chainID: wasmlib.ScChainID, hContract: wasmlib.ScHname, hFunction: wasmlib.ScHname, args: u8[]): u8[];

    Err(): isc.Error;

    postRequest(chainID: wasmlib.ScChainID, hContract: wasmlib.ScHname, hFunction: wasmlib.ScHname, args: u8[], allowance: wasmlib.ScAssets, keyPair: isc.KeyPair, nonce: u64): wasmlib.ScRequestID;

    subscribeEvents(msg: /* chan */ string[], done: /* chan */ bool): isc.Error;

    waitUntilRequestProcessed(chainID: wasmlib.ScChainID, reqID: wasmlib.ScRequestID, timeout: u32): isc.Error;
}

export class WasmClientService implements IClientService {
    waspClient: isc.WaspClient;
    lastError: isc.Error = null;
    eventPort: string;

    public constructor(waspAPI: string, eventPort: string) {
        this.waspClient = new isc.WaspClient(waspAPI);
        this.eventPort = eventPort;
    }

    public static DefaultWasmClientService(): WasmClientService {
        return new WasmClientService("127.0.0.1:19090", "127.0.0.1:15550");
    }

    public callViewByHname(chainID: wasmlib.ScChainID, hContract: wasmlib.ScHname, hFunction: wasmlib.ScHname, args: u8[]): u8[] {
        let res = this.waspClient.callViewByHname(chainID, hContract, hFunction, args);
        this.lastError = this.waspClient.Err;
        if (this.lastError != null) {
            return [];
        }
        return res;
    }

    public Err(): isc.Error {
        return this.lastError;
    }

    public postRequest(chainID: wasmlib.ScChainID, hContract: wasmlib.ScHname, hFunction: wasmlib.ScHname, args: u8[], allowance: wasmlib.ScAssets, keyPair: isc.KeyPair, nonce: u64): wasmlib.ScRequestID {
        let req = new isc.OffLedgerRequest(chainID, hContract, hFunction, args, nonce);
        req.withAllowance(allowance);
        let signed = req.sign(keyPair);
        this.lastError = this.waspClient.postOffLedgerRequest(chainID, signed);
        if (this.lastError != null) {
            return wasmlib.requestIDFromBytes(null);
        }
        return signed.ID();
    }

    public subscribeEvents(msg: /* chan */ string[], done: /* chan */ bool): isc.Error {
        //TODO
        // return subscribe.Subscribe(this.eventPort, msg, done, false, "");
        return null;
    }

    public waitUntilRequestProcessed(chainID: wasmlib.ScChainID, reqID: wasmlib.ScRequestID, timeout: u32): isc.Error {
        return this.waspClient.waitUntilRequestProcessed(chainID, reqID, timeout);
    }
}
